### PR TITLE
limit the depth of nested bool queries

### LINKED ISSUE #55303: Limit the depth of nested queries
Today we rely on recursion to parse, rewrite and create query builders that contains nested queries. This recursion can lead to stack overflow errors, for instance when dealing with a `bool` query that contains a chain of nested `bool` queries:
````
{
	"bool": {
		"should": {
			"boold": {
				"should": {
					"bool": {
						"should": {
							"bool": {
								"should": {
									...
````
These queries,  usually created by mistake by a script,  can have a negative impact on the cluster if they are not detected early. 
We should have a mechanism to detect them and throw a comprehensive error when we see one in search or when trying to save a query.

### PR DESCRIPTION
**Pull request by chengyang14**
limit the depth of nested bool queries

fix #55303

### CODE REVIEW

**Comment by tvernum** [comment]
@elasticmachine test this please

**Comment by chengyang14** [comment]
Hi @jimczi ,

can you help to review this PR?

**Comment by mayya-sharipova** [comment]
@chengyang14 Thanks for your PR and contribution.  I haven't looked into details myself, but I was wondering if you explored if it would be possible to introduce this check earlier before creating all query builders? Can we do the check during a parsing stage, when we parse a boolean query builder from json? For example, can we  add `nestedDepth` to `AbstractQueryBuilder::parseInnerQueryBuilder?

**Comment by chengyang14** [comment]
@mayya-sharipova , thanks for your suggestion,  I will put this check in the parsing stage. 
I am not familiar with using `XContentParser` to parse json. I need to study how to parse `nestedDepth` as a context in` parser.namedObject(QueryBuilder.class, queryName, context)`

**Comment by chengyang14** [comment]
Hi @mayya-sharipova  ,I have finished putting the nested depth check in the parsing stage, please help to check again.

**Review by mayya-sharipova — COMMENTED** [review]
@chengyang14 Thanks for your PR, the code overall look good to me. I've just added small comments around wording and naming.

This PR is still missing:
- tests.  We can add a new test in `BoolQueryBuilderTests.java` where we try to parse a bool query with a big nested depth and expect it to fail. Look at the test `testEmptyBooleanQuery` as an example how a query gets parsed in a test
- documentation. We need to document a new setting for max nested depth for bool queries. We can add this setting into docs/reference/modules/indices/search-settings.asciidoc

File: docs/reference/modules/indices/search-settings.asciidoc [hunk 1]
```diff
@@ -24,4 +24,11 @@ few resources.
...
-Requests that attempt to return more than this limit will return an error.
\ No newline at end of file
+Requests that attempt to return more than this limit will return an error.
+
+[[indices-query-bool-max-nested-depth]]
+`indices.query.bool.max_nested_depth`::
+(<<static-cluster-setting,Static>>, integer) Maximum nested depth of bool queries. Defaults to `20`.
++
+This setting limits the nesting depth of bool queries. Nesting too deep will lead to stack overflow
+errors.
```
Comment (mayya-sharipova) [inline docs/reference/modules/indices/search-settings.asciidoc:34]: May be a little paraphrase the last sentence: "Deep nesting of boolean queries may lead to stack overflow."

File: server/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTests.java [hunk 2]
```diff
@@ -447,4 +450,17 @@ public void testMustRewrite() throws IOException {
...
+    public void testExceedMaxNestedDepth() throws IOException {
+        BoolQueryBuilder query = new BoolQueryBuilder();
+        query.should(new BoolQueryBuilder().should(new BoolQueryBuilder().should(RandomQueryBuilder.createQuery(random()))));
+        BoolQueryBuilder.setMaxNestedDepth(2);
+        try (XContentParser parser = createParser(JsonXContent.jsonXContent, query.toString())) {
+            XContentParseException e = expectThrows(XContentParseException.class,
+                () -> parseInnerQueryBuilder(parser));
+            assertThat(e.getCause().getCause(), Matchers.instanceOf(IllegalArgumentException.class));
+            assertEquals(e.getCause().getCause().getMessage(), "The nested depth of the query exceeds the maximum nested depth for bool queries set in [" +
+                INDICES_MAX_NESTED_DEPTH_SETTING.getKey() + "]");
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTests.java:463]: The format of `assertEquals` is `assertEquals(expected, actual)`, so we should write:
```java
assertEquals("The nested...", e.getCause().getCause().getMessage());
```

**Review by mayya-sharipova — APPROVED** [review]
@chengyang14 Thanks a lot, the PR overall looks very good to me.
I've just left two tiny comments. Please address them, and after that I will run CI tests on this PR.

**Comment by chengyang14** [comment]
@mayya-sharipova Thanks for your review and suggestions.

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by mayya-sharipova** [comment]
@chengyang14 Thanks for your work, your PR has been merged to master. Would you like to create the same PR against 7.x branch, so that we can backport your change to v 7.12? If you don't have time, no worries, I can do the backport myself.

**Comment by chengyang14** [comment]
@mayya-sharipova  I will create the same PR for the 7.x branch.

**Pull request merged**